File: c++/93. Restore IP Addresses.cpp
class Solution {
public:
    vector<string> restoreIpAddresses(string s) {
        vector<string> res;
        for (int i = 1; i <= 3; i++)
            dfs(0, 0, i, "", s, res);
        return res;
    }
    
    void dfs(int cnt, int start, int end, string str, string &s, vector<string> &res) {
        if (end > s.size() || cnt > 3) return;
        string sub = s.substr(start, end - start);
        if (sub.size() > 1 && sub[0] == '0') return;
        if (atoi(sub.c_str()) > 255) return;
        str += sub;
        if (end != s.size())
            str += '.';
        else {
            if (cnt == 3) {
                res.push_back(str);
                return;
            }
        }
        
        for (int i = 1; i <= 3; i++)
            dfs(cnt+1, end, end + i, str, s, res);
    }
    
};

File: python/93. Restore IP Addresses.py
class Solution(object):
    def restoreIpAddresses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        res = []

        def dfs(start, end, path):
            if end > len(s) or len(path)>4: return
            sub = s[start:end]
            if len(sub)>1 and sub[0]=='0': return
            if int(sub)>255: return 
            path.append(sub)
            if end==len(s) and len(path)==4:
                res.append('.'.join(path))
                return
            
            for i in range(1, 4):
                dfs(end, end+i, path[:])

        for i in range(1, 4):
            dfs(0, i, [])
        return res

# 思路：定义start和end两个索引来进行dfs        